cmd/gardener-admission-controller/app/gardener_admission_controller.go

```diff
@@ -16,222 +16,197 @@ package app
-	configObj, gvk, err := o.codecs.UniversalDecoder().Decode(data, nil, nil)
-	if err != nil {
-		return nil, err
-	}
-	config, ok := configObj.(*config.AdmissionControllerConfiguration)
-	if !ok {
-		return nil, fmt.Errorf("got unexpected config type: %v", gvk)
+func (o *options) validate() error {
+	if errs := configvalidation.ValidateAdmissionControllerConfiguration(o.config); len(errs) > 0 {
+		return errs.ToAggregate()
 	}
-	return config, nil
-}
 
-// AdmissionController contains all necessary information to run the admission controller.
-type AdmissionController struct {
-	Config    *config.AdmissionControllerConfiguration
-	ClientMap clientmap.ClientMap
+	return nil
 }
 
+// run runs gardener-admission-controller using the specified options.
 func (o *options) run(ctx context.Context) error {
-	c, err := o.loadConfigFromFile(o.configFile)
-	if err != nil {
-		return err
-	}
+	log := logf.Log
 
-	if errs := configvalidation.ValidateAdmissionControllerConfiguration(c); len(errs) > 0 {
-		return errs.ToAggregate()
+	log.Info("getting rest config")
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		o.config.GardenClientConnection.Kubeconfig = kubeconfig
 	}
 
-	admissionController, err := NewAdmissionController(c)
+	restConfig, err := kubernetes.RESTConfigFromClientConnectionConfiguration(&o.config.GardenClientConnection, nil)
 	if err != nil {
 		return err
 	}
 
-	return admissionController.Run(ctx)
-}
-
-// NewAdmissionController creates a new AdmissionController instance.
-func NewAdmissionController(cfg *config.AdmissionControllerConfiguration) (*AdmissionController, error) {
-	if cfg == nil {
-		return nil, errors.New("config is required")
+	log.Info("setting up manager")
+	mgr, err := manager.New(restConfig, manager.Options{
+		Scheme:                  kubernetes.GardenScheme,
+		LeaderElection:          false,
+		HealthProbeBindAddress:  fmt.Sprintf("%s:%d", o.config.Server.HealthProbes.BindAddress, o.config.Server.HealthProbes.Port),
+		MetricsBindAddress:      fmt.Sprintf("%s:%d", o.config.Server.Metrics.BindAddress, o.config.Server.Metrics.Port),
+		Host:                    o.config.Server.HTTPS.BindAddress,
+		Port:                    o.config.Server.HTTPS.Port,
+		CertDir:                 o.config.Server.HTTPS.TLS.ServerCertDir,
+		GracefulShutdownTimeout: &gracefulShutdownTimeout,
+		Logger:                  log,
+	})
+	if err != nil {
+		return err
 	}
 
-	// Initialize logger
-	logger := logger.NewLogger(cfg.LogLevel)
-	logger.Info("Starting Gardener admission controller...")
-
-	// Prepare a Kubernetes client object for the Garden cluster which contains all the Clientsets
-	// that can be used to access the Kubernetes API.
-	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
-		cfg.GardenClientConnection.Kubeconfig = kubeconfig
+	log.Info("setting up healthcheck endpoints")
+	if err := mgr.AddReadyzCheck("ping", healthz.Ping); err != nil {
```

Comment:
Can you elaborate why we need two checks for `/readyz`?

Reply:
Well, that's a good point. Removed the ping `readyz` check.